### PR TITLE
Fixes #2651 SG gets killed due to excessive memory usage with continu…

### LINKED ISSUE #2651: SG gets killed due to excessive memory usage with continuous doc update
### Sync Gateway version
1.5.0-429

### Operating system
CentOS 7

### Config file
```
{
    "interface":":4984",
    "adminInterface": "0.0.0.0:4985",
    "maxIncomingConnections": 0,
    "maxCouchbaseConnections": 16,
    "maxFileDescriptors": 90000,
    "slowServerCallWarningThreshold": 500,
    "compressResponses": false,
    "log": ["CRUD+", "Cache+", "HTTP+", "Changes+"],
    "verbose":"true",
    "databases":{
        "db":{
	    "unsupported": {
        	"enable_extended_attributes": true
	     },
            "server":"http://<CBS_IP>:8091",
	    "revs_limit": 10000000,
            "bucket":"data-bucket",
            "username":"data-bucket",
            "password": "password",
	    "users": {"GUEST": {"disabled": false,"admin_channels": ["*"]}}            
        }
    }
```
### Steps to reproduce

The aim of this testing was to hit the xattr memory limit and test SG's behavior. I manually tried out the below 2 scenarios. It seems like SG will get killed well before hitting the xattr memory limit.

## scenario 1:
- SG 1.5 without xattrs running on a VM.
- VM has 2GB RAM.
- Add one doc to SG.
- revs_limit is set to 10 million.
- Constantly update the doc with a script.
- After 8535 updates, SG gets killed by the OS.
- When SG gets killed, SG is using around 90% cpu and all the memory and the OS is running out of memory.

/var/log/messages:
```
Jun 14 21:59:16 localhost kernel: 38610 total pagecache pages
Jun 14 21:59:16 localhost kernel: 35800 pages in swap cache
Jun 14 21:59:16 localhost kernel: Swap cache stats: add 16917332, delete 16881532, find 5161984/6611476
Jun 14 21:59:16 localhost kernel: Free swap  = 0kB
Jun 14 21:59:16 localhost kernel: Total swap = 1572860kB
Jun 14 21:59:16 localhost kernel: 524174 pages RAM
Jun 14 21:59:16 localhost kernel: 0 pages HighMem/MovableOnly
Jun 14 21:59:16 localhost kernel: 53130 pages reserved
Jun 14 21:59:16 localhost kernel: [ pid ]   uid  tgid total_vm      rss nr_ptes swapents oom_score_adj name
Jun 14 21:59:16 localhost kernel: [ 3288]     0  3288   831901   398350    1519   366000             0 sync_gateway
Jun 14 21:40:02 localhost kernel: Out of memory: Kill process 2777 (sync_gateway) score 874 or sacrifice child
Jun 14 21:40:02 localhost kernel: Killed process 2777 (sync_gateway) total-vm:3285600kB, anon-rss:1554664kB, file-rss:0kB
```
## scenario 2:

- SG 1.5 with xattrs enabled running on a VM.
- VM has 6GB RAM.
- Add one doc to SG.
- revs_limit is set to 10 million.
- Constantly update the doc with a script.
- After 16568 updates, SG gets killed by the OS.
- When SG gets killed, SG is using around 90% cpu and all the memory and the OS is running out of memory.

/var/log/messages:
```
Jun 14 22:53:19 localhost kernel: Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=2048kB
Jun 14 22:53:19 localhost kernel: 62698 total pagecache pages
Jun 14 22:53:19 localhost kernel: 58157 pages in swap cache
Jun 14 22:53:19 localhost kernel: Swap cache stats: add 2824375, delete 2766218, find 763838/1027429
Jun 14 22:53:19 localhost kernel: Free swap  = 0kB
Jun 14 22:53:19 localhost kernel: Total swap = 1572860kB
Jun 14 22:53:19 localhost kernel: 1572750 pages RAM
Jun 14 22:53:19 localhost kernel: 0 pages HighMem/MovableOnly
Jun 14 22:53:19 localhost kernel: 86030 pages reserved
Jun 14 22:53:19 localhost kernel: [ pid ]   uid  tgid total_vm      rss nr_ptes swapents oom_score_adj name
Jun 14 22:53:20 localhost kernel: [ 2368]     0  2368  1796765  1370579    3421   367426             0 sync_gateway
Jun 14 22:53:20 localhost kernel: Out of memory: Kill process 2368 (sync_gateway) score 898 or sacrifice child
Jun 14 22:53:20 localhost kernel: Killed process 2368 (sync_gateway) total-vm:7187060kB, anon-rss:5482280kB, file-rss:36kB
```

### Expected behavior
SG memory usage should be constant.

### Actual behavior
SG memory usage grows constantly.

### PR DESCRIPTION
**Pull request by tleyden**
…ous doc update

Fixes #2651 

Notes:

- I left out the change to convert fmt.Sscanf() as mentioned in [this comment](https://github.com/couchbase/sync_gateway/issues/2651#issuecomment-309616077), because it seemed like a red herring.
- I put the `RenderGraphvizDot()` change in a [separate PR](https://github.com/couchbase/sync_gateway/pull/2670)

Passing unit tests against couchbase bucket jobs:

- [xattrs=true](http://uberjenkins.sc.couchbase.com/view/Build/job/build-sync-gateway/117/parameters/)
- [xattrs=false](http://uberjenkins.sc.couchbase.com/view/Build/job/build-sync-gateway/115/parameters/)

### CODE REVIEW

File: db/revtree.go [hunk 1]
```diff
@@ -301,51 +326,180 @@ func (tree RevTree) copy() RevTree {
...
 		}
 	}
 	return
+
+}
+
+// Find the minimum generation that has a non-deleted leaf.  For example in this rev tree:
+//   http://cbmobile-bucket.s3.amazonaws.com/diagrams/example-sync-gateway-revtrees/three_branches.png
+// The minimim generation that has a non-deleted leaf is "7-non-winning unresolved"
+func (tree RevTree) GenerationShortestNonTombstonedBranch() int {
```
Comment (adamcfraser) [inline db/revtree.go]: The naming of these methods is hard to parse - it's hard to tell whether Generation is supposed to be a noun or a verb. Maybe 'CalculateShortest...' or something?

**Review by adamcfraser — CHANGES_REQUESTED** [review]
Functionally it looks good - some style questions, and then a question about performance.

File: db/revtree.go [hunk 2]
```diff
@@ -162,17 +167,37 @@ func (tree RevTree) getHistory(revid string) []string {
 
 // Returns the leaf revision IDs (those that have no children.)
 func (tree RevTree) GetLeaves() []string {
+	acceptAllLeavesFilter := func(revId string) bool {
+		return true
+	}
+	return tree.GetLeavesFiltered(acceptAllLeavesFilter)
+}
+
+func (tree RevTree) GetTombstonedLeaves() []string {
```
[the remark this answers is not in the document]
Comment (tleyden) [reply db/revtree.go]: Yes, it ended up being test-only.  I'll move it.

File: db/revtree.go [hunk 3]
```diff
@@ -301,51 +326,180 @@ func (tree RevTree) copy() RevTree {
...
+// - Find the generation of the shortest non-tombstoned branch (eg, 100)
+// - Calculate the tombstone generation threshold based on this formula:
+//      tombstoneGenerationThreshold = genShortestNonTSBranch - maxDepth
+//      Ex: if maxDepth is 20, and tombstoneGenerationThreshold is 100, then tombstoneGenerationThreshold will be 80
+// - Check each tombstoned branch, and if the leaf node on that branch has a generation older (less) than
+//   tombstoneGenerationThreshold, then remove all nodes on that branch up to the root of the branch.
 func (tree RevTree) pruneRevisions(maxDepth uint32, keepRev string) (pruned int) {
-	if len(tree) <= int(maxDepth) {
-		return 0
+
```
[the remark this answers is not in the document]
Comment (tleyden) [reply db/revtree.go]: Agreed, fixed in a32aee915ffc103c98099bafd4126b9a82861a5d

File: db/revtree.go (hunk 1 above)
Comment (tleyden) [reply db/revtree.go]: Fixed in 29a496103dc8579c512960be41a2895fe68d004c

Renamed to:

```
func (tree RevTree) FindShortestNonTombstonedBranch() (generation int) {
    ...
}
```

File: db/revtree.go [hunk 4]
```diff
@@ -301,51 +326,180 @@ func (tree RevTree) copy() RevTree {
...
+				genLongestTSBranch = gen
+			}
+		}
+	}
+	return genLongestTSBranch
+}
+
+// Find the length of the longest branch
+func (tree RevTree) LongestBranch() int {
+
```
[the remark this answers is not in the document]
Comment (tleyden) [reply db/revtree.go]: Fixed in 29a496103dc8579c512960be41a2895fe68d004c

**Review by adamcfraser — APPROVED** [review]

**Comment by curiousily** [comment]
Hi @tleyden,

Do you have any idea if this issue affects SG 1.4.1 as well?

Thank you for fixing this.

**Comment by tleyden** [comment]
@curiousily yes, it definitely affects SG 1.4.1.

**Comment by curiousily** [comment]
@tleyden any idea if it is going to be in SG 1.4.2 or something? We might be facing this issue (or something similar) in production. Thanks!

**Comment by tleyden** [comment]
It's still under discussion, but I will update this ticket as soon as I know more.  Also, I'll let folks know that users are looking for a 1.4.x patch.

**Comment by curiousily** [comment]
Thanks! Will be waiting for updates.
